docs/proposals/cascading-deletion.md

```diff
@@ -0,0 +1,235 @@
+  * Update handler:
+    * If the update removes the last finalizer, and the DeletionTimestamp is non-nil, and the DeletionGracePeriodSeconds is zero, then deletes the object from the registry.
+    * If the update removes the last finalizer, and the DeletionTimestamp is non-nil, but the DeletionGracePeriodSeconds is non-zero, then just updates the object.
+
+## 2B. Enforcing orphaning policty with a finalizer
+
+## API changes
+
+```
+type ObjectMeta struct {
+	...
+	ParentReferences []ObjectReference
+}
+```
+
+**ObjectMeta.ParentReferences**: same as in [design #1](#api-changes).
+
+## New components
+
+* Garbage Collector: Almost identical to the [Garbage Collector](#new-components) in design #1. Some notable differences are
+  * The Garbage Processor doesn't need to remove dangling ParentReferences. The "orphan" finalizer takes care of that.
+  * The Garbage Processor sends a deletion request for an item if none of the object listed in the ParentReferences exist.
+  * Adding a fourth component, the"orphan" finalizer:
+    * Watches for update events as described in [2A](#new-components-1).
+    * Removes the object in the event from the `ParentReferences` of its children.
+      * children can be found via the DAG kept by the GC, or by relisting the child resource and checking the ParentReferences field of each potential child object.
+    * Removes the finalizer from the `ObjectMeta.Finalizers` of this object.
+
+# Comparison between design #1 and #2
+
+The main difference between the two designs is when to update the ParentReferences. In design #1, because a tombstone is created to indicate "orphaning" is desired, the updates to ParentReferences can be deferred until the deletion of the tombstone. In design #2, the updates need to be done before the parent object is deleted from the registry.
+
+* Advantages of "Tombstone + GC" design
+  * Faster to free the resource name compared to using finalizers. The original object can be deleted to free the resource name once the tombstone is created, rather than waiting for the finalizers to update all children’s ObejectMeta.ParentReferences.
+* Advantages of "Finalizer Framework + GC"
+  * The finalizer framework is needed for other purposes as well.
+
+# Common designs
+
+## Orphan adoption
+
+Controllers are responsible for adopting orphaned child resources. To do so, controllers
+* Checks a potential child object’s ParentReferences to determine if it is orphaned.
+* Fills the ParentReferences if the object matches the controller’s selector and is orphaned.
+
+There is a potential race between the GC removing a parent reference and the controllers adding it back during adoption. Imaging this case: a user deletes a parent object and intends to orphan the child objects, so the GC removes the parent from the child object's ParentReferences list, but the controller of the parent resource hasn't observed the deletion yet, so it adopts the child again and adds the reference back.
```

Comment:
There are no cases of multiple controllers acting on the same API object, are there?